### DISCUSSION TITLE
Why are filters not supported in Firefox?

### DISCUSSION THREAD

**Discussion by MLNW** [skeletonlabs/skeleton#1426] [Q&A] [answered]
If I understand correctly then Firefox **does** support filters according to the link in [the documentation](https://www.skeleton.dev/actions/filters#browser-support): https://caniuse.com/css-filters

![grafik](https://user-images.githubusercontent.com/20930814/236235542-2d1ce737-28de-4bee-af73-4ac27062ddf2.png)

So how come [filters](https://www.skeleton.dev/actions/filters) are not supported in Firefox?

**Comment by Sarenor** [comment]
I can't find the previous issue or discussion about this, basically because they theoretically work but not how Skeleton does it and Chris hasn't found a way (or the time to try to find another way) to make them work with Firefox.

You can try yourself though, check out Skeleton, disable the "don't run this in Firefox" check and see if it works now. You may bring us good news!

**Reply by MLNW** [reply]
I did find the other thread: https://github.com/skeletonlabs/skeleton/discussions/1180

It seems like I didn't find it because by default only open discussions are considered for searches ...

**Comment by endigo9740** [comment]
The answer is simple - try them. They don't work. Doesn't matter what that page says if it doesn't work in the real world :)

**Reply by MLNW** [reply]
You are right that the filters do not work on the Skeleton website. The filters on these websites do work in Firefox though:
* https://developer.mozilla.org/en-US/docs/Web/CSS/filter
* https://developer.mozilla.org/en-US/docs/Web/CSS/Filter_Effects

Maybe its fundamentally something different, but it sounds to me like it shouldn't be any different.

**Reply by endigo9740** [reply, marked as answer]
Filter is not a singular feature but rather a set - including blur, drop shadow, etc. The one we need specifically is `filter: url()`, which applies an visual filter effects defined in an SVG format. Here's what one of the filters looks like:

```js
<!-- Apollo: `filter: url(#Apollo)` -->
<svg id="svg-filter-apollo" class="hidden">
	<filter id="Apollo" filterUnits="objectBoundingBox" primitiveUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
		<feColorMatrix
			values="0.8 0.6 -0.4 0.1 0,
					0 1.2 0.05 0 0,
					0 -1 3 0.02 0,
					0 0 0 50 0"
			result="final"
			in="SourceGraphic"
		/>
	</filter>
</svg>
```

The Skeleton action for applying filters in Skeleton will never work in Firefox as is, as we exclude it explicitly. There's no sense running logic that won't visually change anything. You an see that here:
https://github.com/skeletonlabs/skeleton/blob/dev/src/lib/actions/Filters/filter.ts#L6

If we disable this exemption and allow the Action to run and apply the filter - this is the result in Firefox though:
![Screenshot 2023-05-04 at 11 05 23 AM](https://user-images.githubusercontent.com/1509726/236266478-6fe420f8-de8c-4603-beb1-c4b105c91d33.png)

Unfortunately this conversation comes up every 2-3 weeks. Someone points to the caniuse site, argues that it should work, then we show them this. I'll offer the same response I do to everyone else - if you can find a way to get it working, send a PR. We would welcome this in a heartbeat!

**Reply by MLNW** [reply]
Thanks for your detailed response about the issue. I understand that the limitation is due to technical constraints rather than choice.

I opened this discussion thread since I couldn't find any related information about filters not working in Firefox in the existing issues or discussions. I believe that having this thread will help future users who might be looking for the same information.

In that sense it might be helpful to update the documentation to mention the incompatibility issue with Firefox more explicitly. This could prevent users from being misled by the current reference to the Can I Use page, which might give the impression that the problem is with Skeleton itself. By providing clearer information about this limitation, users may better understand the current state of compatibility and potentially encourage contributions to improve the library's compatibility with Firefox.

Thanks again for your time and effort in maintaining this library :)

**Reply by endigo9740** [reply]
Yeah I believe most discussion around this has happened on Discord, so I think this will be useful. We'll point folks here going forward. I agree on the suggestion for the doc updates too. Best to centralize that information.

Fingers crossed Firefox does eventually support this, as I'd love to remove the exclusion!

**Reply by MLNW** [reply]
I had a short look around the web to find any reference of the issue but wasn't able to find much. The MDN documentation only states that it should work. It almost sounds like there is no awareness of the issue, in which case it won't be fixed any time soon.

Were you able to find references to the issue online? If not it would be best to raise awareness, e.g., through an issue on their bug tracker.

**Reply by endigo9740** [reply]
Unfortunately I haven't had a chance to dig into this. I'd welcome you filing an issue on our behalf if you get a chance @MLNW

**Reply by endigo9740** [reply]
FYI: https://github.com/skeletonlabs/skeleton/issues/1431

**Reply by MLNW** [reply]
I created an MWE and raised an issue in BugZilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1831771

Let's see where this goes.

**Reply by MLNW** [reply]
As part of the issue I created a fully integrated example that does not depend on Svelte or Vite and in that configuration got filters to work in Firefox. Based on that I was able to get my initial MWE to also render filters. It seems like the SVG containing the filter was removed by Firefox in that initial example.

With that I think it is confirmed that the issue does not lie with Firefox directly but somewhere between Firefox, Vite/Svelte, and Skeleton. In that case it should be possible for us to fix the issue on our side instead of having to rely on a fix coming from Firefox.

I think we need to have another look at this and discuss possible solutions to the issue. Let me know if you are up for that @endigo9740

**Reply by MLNW** [reply]
I actually got it to work by simply removing the `hidden` class from the SVGs in the DOM:

![grafik](https://user-images.githubusercontent.com/20930814/236678115-29dd7feb-1061-42f1-b57e-956291a6199c.png)

This is obviously only a proof of concept but it shows that it is possible to fix on our side.

**Reply by MLNW** [reply]
I have created a draft PR: https://github.com/skeletonlabs/skeleton/pull/1449

Let's continue implementation discussion there.

**Reply by endigo9740** [reply]
I'm away for the weekend but will review this tomorrow. Thanks @MLNW

**Reply by MLNW** [reply]
Solved via https://github.com/skeletonlabs/skeleton/pull/1449
